index.js

```diff
@@ -0,0 +1,865 @@
+                 String(other) === String(this);
+        },
+
+        //  String#concat :: String ~> String -> String
+        'fantasy-land/concat': function(s) {
+          return this + s;
+        }
+      }
+    },
+
+    Array: {
+      //  Array.empty :: () -> Array a
+      'fantasy-land/empty': function() { return []; },
+
+      //  Array.of :: a -> Array a
+      'fantasy-land/of': function(x) { return [x]; },
+
+      //  Array.chainRec :: ((a -> c, b -> c, a) -> Array c, a) -> Array b
+      'fantasy-land/chainRec': function(f, x) {
+        var next = function(x) { return {value: x, done: false}; };
+        var done = function(x) { return {value: x, done: true}; };
+        var $todo = [x];
+        var $done = [];
+        while ($todo.length > 0) {
+          var xs = f(next, done, $todo.shift());
+          var $more = [];
+          for (var idx = 0; idx < xs.length; idx += 1) {
+            (xs[idx].done ? $done : $more).push(xs[idx].value);
+          }
+          Array.prototype.unshift.apply($todo, $more);
+        }
+        return $done;
+      },
+
+      prototype: {
+        //  Array#toString :: Array a ~> () -> String
+        toString: function() {
+          return '[' + this.map(toString).join(', ') + ']';
+        },
+
+        //  Array#equals :: Array a ~> Array a -> Boolean
+        'fantasy-land/equals': function(other) {
+          if (other.length !== this.length) return false;
+          for (var idx = 0; idx < this.length; idx += 1) {
+            if (!equals(other[idx], this[idx])) return false;
+          }
+          return true;
+        },
+
+        //  Array#concat :: Array a ~> Array a -> Array a
+        'fantasy-land/concat': function(xs) { return this.concat(xs); },
+
+        //  Array#map :: Array a ~> (a -> b) -> Array b
+        'fantasy-land/map': function(f) {
+          return this.map(function(x) { return f(x); });
+        },
+
+        //  Array#ap :: Array a ~> Array (a -> b) -> Array b
+        'fantasy-land/ap': function(fs) {
+          var xs = this;
+          return chain(function(f) { return map(f, xs); }, fs);
+        },
+
+        //  Array#chain :: Array a ~> (a -> Array b) -> Array b
+        'fantasy-land/chain': function(f) {
+          var result = [];
+          this.forEach(function(x) {
+            Array.prototype.push.apply(result, f(x));
+          });
+          return result;
+        },
+
+        //  Array#reduce :: Array a ~> ((b, a) -> b, b) -> b
+        'fantasy-land/reduce': function(f, initial) {
+          return this.reduce(function(acc, x) { return f(acc, x); }, initial);
+        },
+
+        //  Array#traverse :: Applicative f => Array a ~> ((a -> f b), (c -> f c)) -> f (Array b)
+        'fantasy-land/traverse': function(f, of) {
+          var applicative = of([]);
+          for (var idx = this.length - 1; idx >= 0; idx -= 1) {
+            applicative = ap(map(prepend, f(this[idx])), applicative);
```

Comment:
The performance actually seems reasonable. I did revert to an imperative implementation of `Array#fantasy-land/ap` which made a noticeable difference. Producing an array containing 10^6 arrays now takes less than one second on my computer. Using linked lists as suggested by @scott-christopher results in the same task taking slightly more than a second for me.

Here's the diff, for reference:

```diff
--- a/index.js
+++ b/index.js
@@ -132,6 +132,25 @@

   'use strict';

+  //  Nil :: List a
+  var Nil = {isNil: true, isCons: false};
+
+  //  Cons :: a -> List a -> List a
+  var Cons = function(head) {
+    return function(tail) {
+      return {isNil: false, isCons: true, head: head, tail: tail};
+    };
+  };
+
+  //  listToArray :: List a -> Array a
+  var listToArray = function(_list) {
+    var result = [];
+    for (var list = _list; list.isCons; list = list.tail) {
+      result.push(list.head);
+    }
+    return result;
+  };
+
   //  has :: (String, StrMap a) -> Boolean
   var has = function(k, strMap) {
     return Object.prototype.hasOwnProperty.call(strMap, k);
@@ -432,11 +451,11 @@

         //  Array#traverse :: Applicative f => Array a ~> ((a -> f b), (c -> f c)) -> f (Array b)
         'fantasy-land/traverse': function(f, of) {
-          var applicative = of([]);
+          var applicative = of(Nil);
           for (var idx = this.length - 1; idx >= 0; idx -= 1) {
-            applicative = ap(map(prepend, f(this[idx])), applicative);
+            applicative = ap(map(Cons, f(this[idx])), applicative);
           }
-          return applicative;
+          return map(listToArray, applicative);
         }
       }
     },
```

This expression produces an array containing 10^6 arrays:

```javascript
Z.traverse(R.of, R.identity, R.repeat(R.range(0, 10), 6))
```

I imagine there's still room for improvement. Feel free to experiment, @svozza. :)